### ISSUE TITLE
 Early convergence Issue

### ISSUE THREAD

**Issue by chuanzihe** [e-lab/ENet-training#8]
Hi,

I found several model I trained would reach the great performance that the model-best.net you offered. I set opt.lua according to your documentation, except that I used a batch size of 2.

Here's my result:
![image](https://cloud.githubusercontent.com/assets/19618625/17373875/02d82c3a-595f-11e6-9e10-063753b9c951.png)

and there is the result tested by your model-best.net

![image](https://cloud.githubusercontent.com/assets/19618625/17374377/2dc527e8-5961-11e6-9d62-a550ccbf6528.png)

I trained it on cityscapes for several time. Actually the training process tended to converge at a early stage(80-100th epoch). Here is a graph of test error trend. 
<img width="595" alt="2016-08-03 9 44 05" src="https://cloud.githubusercontent.com/assets/19618625/17373840/d6186f52-595e-11e6-810e-e546ba7b172c.png">

For parameter settings, I basically followed your default setting or your documentation. 

**Encoder**

```
smallNet : false
learningRate : 0.001
datahistClasses :   810274
 1969251
  321342
 1195839
   34488
   46349
   65294
   10960
   29268
  835925
   61092
  202450
   65301
    7305
  371063
   14192
   12671
   12402
    5239
   22095
[torch.FloatTensor of size 20]

batchSize : 2
dataconClasses : table: 0x401fe608
dataClasses : table: 0x401fe4d0
channels : 3
printNorm : false
save : savemodel/
CNNEncoder : historymodel/enc_1/model-best.net
labelHeight : 32
labelWidth : 64
plot : false
nGPU : 2
lrDecayEvery : 100
weightDecay : 0.0005
imHeight : 256
dataset : cs
momentum : 0.9
devid : 1
cachepath : historymodel/
datapath : datasets/Cityscapes/
threads : 8
maxepoch : 300
noConfusion : all
learningRateDecay : 1e-07
model : models/encoder.lua
imWidth : 512

```

and I got `Best test error: 0.46744307547808, in epoch: 88`

**Decoder**

```
smallNet : false
learningRate : 0.001
datahistClasses :   45323724
 127228632
  20988628
  78573240
   2259981
   3025684
   4230457
    716094
   1900118
  54828056
   3993315
  13767863
   4206164
    466290
  24125912
    922413
    811419
    803171
    340114
   1427811
[torch.FloatTensor of size 20]

batchSize : 2
dataconClasses : table: 0x400dd688
dataClasses : table: 0x400dd550
channels : 3
printNorm : false
save : savemodel/
CNNEncoder : historymodel/enc_2_728/model-best.net
labelHeight : 256
labelWidth : 512
plot : false
nGPU : 2
lrDecayEvery : 100
weightDecay : 0.0005
imHeight : 256
dataset : cs
momentum : 0.9
devid : 1
cachepath : historymodel/dec
datapath : /home/eeb433/Documents/Yuhang/dilation/datasets/Cityscapes/
threads : 8
maxepoch : 300
noConfusion : all
learningRateDecay : 1e-07
model : models/decoder.lua
imWidth : 512

```

and I got `Best test error: 0.77709275662899, in epoch: 95`

Thanks for your answer!

**Comment by chuanzihe** [comment]
Your paper says

> Learning rate of 5e−4 and L2 weight decay of 2e−4, along with batch size of 10 consistently provided the best results.

However even if I adjusted the learning rate and L2 weight decay to train on **CamVid**, that result was also terrible, with 

**encoder** `Best test error: 0.51656901772166, in epoch: 36`

and

**decoder** `Best test error: 0.87722304180778, in epoch: 98`

For your information, 

**Encoder**

```
1 : models/encoder.lua
smallNet : false
learningRate : 0.0005
datahistClasses :   38291
 172613
 231872
   8943
 304205
  44557
  97849
  11778
  11186
  57618
   6353
   2935
[torch.FloatTensor of size 12]

batchSize : 5
dataconClasses : table: 0x408e3ff8
dataClasses : table: 0x408e3f40
channels : 3
model : models/encoder.lua
save : savemodel/
CNNEncoder : historymodel/enc_2_728/model-best.net
labelHeight : 45
labelWidth : 60
printNorm : false
nGPU : 2
lrDecayEvery : 100
weightDecay : 0.0002
imHeight : 360
dataset : cv
momentum : 0.9
devid : 1
cachepath : historymodel/enc_cv/
datapath : Data/CamVid/
threads : 8
maxepoch : 300
noConfusion : all
learningRateDecay : 1e-07
plot : false
imWidth : 480
```

**Decoder**

```
1 : models/decoder.lua
smallNet : false
learningRate : 0.0005
datahistClasses :   2496015
 10659041
 14685353
   621445
 20060738
  2833188
  6164459
   741316
   714595
  3679026
   404654
   184967
[torch.FloatTensor of size 12]

batchSize : 5
dataconClasses : table: 0x419f9d70
dataClasses : table: 0x419f9cb8
channels : 3
printNorm : false
save : savemodel/
CNNEncoder : historymodel/enc_cv_1/model-best.net
labelHeight : 360
labelWidth : 480
weightDecay : 0.0002
nGPU : 2
lrDecayEvery : 100
model : models/decoder.lua
imHeight : 360
dataset : cv
momentum : 0.9
devid : 1
cachepath : historymodel/enc_cv_de/
datapath : Data/CamVid/
threads : 8
maxepoch : 300
noConfusion : all
learningRateDecay : 1e-07
plot : false
imWidth : 480
```

Thank you!

**Comment by ghost** [comment]
If  I were you, I would decrease the learning rate. This will make the training converge slower but surely. Btw, did you draw the error curve online or offline ,using what software ? ;)

**Comment by codeAC29** [comment]
batchSize affects the result. I see that you have used a batchSize of 2. Try using a batchSize of 8/10.
For us also network converged before 100 epochs.

**Comment by ghost** [comment]
@codeAC29 Are you saying that increasing the batch size does improve the accuracy ? If so, why is that ? Any intuition behind it ? thx.

**Comment by chuanzihe** [comment]
@codeAC29 Thank you Abhishek! Since I only have 1 GPU so I could only afford a batch size of 2. Maybe I could try to change the iteration size to achieve same result.

**Comment by chuanzihe** [comment]
@djebm2 I simply use keynote to plot this.

**Comment by codeAC29** [comment]
You can use `plot` option too while training if you want the plots. We are using gnu-plot for this.

**Comment by Mina1368** [comment]
Hi 
I have the same problem with CamVid dataset. I also used batchSize  = 2
Did you find anything useful?

Thanks,
Mina

**Comment by ghost** [comment]
Is there any update on this batch size and learning rate selection? We only have one GPU, so the best batch size for CamVid I can do is 5, most time is 3 or 4. To adjust learning rate, should I reduce the learning rate to a smaller number or to decay it more frequently?

**Comment by codeAC29** [comment]
@wzhouuu look into #15

**Comment by korkinof** [comment]
Small batch size can lead to very noisy gradients, so I would not be surprised if that is one of the issues.
Btw, has anyone made any progress with this issue or a Caffe implementation?
Best,
D.

**Comment by ghost** [comment]
This is my opinion after training a lot of models, correct me if I'm wrong.
The batch size significantly affects where you can get the best result and how long you need to train the model. I saw @clairehe decayed the learning rate every 100 epochs, which is actually a big issue. You probably only got the local minimum. Try decay it after 300 or even more epochs, you'll find the best model actually appears later. The smaller batch size you use, the longer training you should expect.
Instead of setting a fixed number to decay learning rate, I modified the code so that if I couldn't get a better results (the error didn't reduce) for 150 epochs (flexible number), decay the learning rate. Say if you get the best model at epoch 88, count from 0, if no more better result occurs after count to 150, decay the learning rate at epoch (88+150=) 238.        Hope this helps.

**Comment by davyneven** [comment]
Did someone managed to reach similar performance, training with a smaller batchsize? I trained with a batchsize of 2, simulating a batchsize of 10 by accumulating the gradients, but can't reach same accuracy as the provided best model.
